### PR TITLE
:ghost: addon adapter using more binding.

### PR DESCRIPTION
**Pull request by jortel**
Refactor _addon_ package to more completely use the _binding_.RichClient.
To accomplish this,  a few API _handlers_ (or methods) copied from _addon_ to _binding_. This also eliminates the need for the handler objects (int binding) to export the `Client` attribute.

Exit code=1 only on _hard_ errors.  Retrying the pod for _soft_ errors is unlikely to have a different result.

### CODE REVIEW

File: addon/adapter.go [hunk 1]
```diff
@@ -79,11 +107,9 @@ func (h *Adapter) Run(addon func() error) {
 		if err != nil {
 			if _, soft := err.(interface{ Soft() *SoftError }); !soft {
 				Log.Error(err, "Addon failed.")
+				os.Exit(1)
 			}
-			if h.client.Error == nil {
-				h.Failed(err.Error())
-			}
-			os.Exit(1)
```
Comment (jortel) [inline addon/adapter.go:86]: **hint**: Exit (1) only for _hard_ errors.

File: addon/adapter.go [hunk 2]
```diff
@@ -79,11 +107,9 @@ func (h *Adapter) Run(addon func() error) {
 		if err != nil {
 			if _, soft := err.(interface{ Soft() *SoftError }); !soft {
 				Log.Error(err, "Addon failed.")
+				os.Exit(1)
 			}
-			if h.client.Error == nil {
-				h.Failed(err.Error())
```
Comment (jortel) [inline addon/adapter.go:84]: **hint**: The client will no-op when an error is set.  No need to check here.

File: binding/application.go [hunk 3]
```diff
@@ -23,44 +29,317 @@ func (h *Application) Create(r *api.Application) (err error) {
...
-	b = &Bucket{
-		path:   path,
-		Client: h.Client,
+	b = &BucketContent{
+		root:   path,
+		client: h.client,
 	}
 	return
 }
+
```
Comment (jortel) [inline binding/application.go:73]: FindIdentity()
Tags()
Facts()

Copied here from addon/

File: binding/bucket.go [hunk 4]
```diff
@@ -1,36 +1,89 @@
...
 
 //
 // Bucket API.
 type Bucket struct {
-	// hub API client.
-	Client *Client
-	// root path
-	path string
+	client *Client
+}
```
Comment (jortel) [inline binding/bucket.go:12]: **hint**: Bucket copied from addon/

File: binding/identity.go [hunk 5]
```diff
@@ -7,45 +7,65 @@ import (
...
 }
 
 //
 // Get a Identity by ID.
 func (h *Identity) Get(id uint) (r *api.Identity, err error) {
 	r = &api.Identity{}
+	p := Param{
+		Key:   api.Decrypted,
+		Value: "1",
+	}
```
Comment (jortel) [inline binding/identity.go:27]: Adding the _decrypted_ parameter.  This assumes that for binding users, there is no use case for fetching encrypted identities.

File: binding/proxy.go [hunk 6]
```diff
@@ -23,29 +22,47 @@ func (h *Proxy) Create(r *api.Proxy) (err error) {
...
 // Delete a Proxy.
 func (h *Proxy) Delete(id uint) (err error) {
-	err = h.Client.Delete(Path(api.ProxyRoot).Inject(Params{api.ID: id}))
+	err = h.client.Delete(Path(api.ProxyRoot).Inject(Params{api.ID: id}))
+	return
+}
+
+//
+// Find by Kind.
+// Returns nil when not found.
```
Comment (jortel) [inline binding/proxy.go:54]: hint: Find() copied from addon/.  Used by the addons and (tackle2-addon) but generally useful enough to be included in the binding.

File: binding/tag.go [hunk 7]
```diff
@@ -23,29 +22,66 @@ func (h *Tag) Create(r *api.Tag) (err error) {
...
 }
 
 //
 // Delete a Tag.
 func (h *Tag) Delete(id uint) (err error) {
-	err = h.Client.Delete(Path(api.TagRoot).Inject(Params{api.ID: id}))
+	err = h.client.Delete(Path(api.TagRoot).Inject(Params{api.ID: id}))
+	return
+}
+
```
Comment (jortel) [inline binding/tag.go:51]: **hint**: Coped from addon/ and generally useful enough for all binding users to be included.
Find()
Ensure()

vvv

File: binding/tagcategory.go [hunk 8]
```diff
@@ -23,29 +22,65 @@ func (h *TagCategory) Create(r *api.TagCategory) (err error) {
...
 }
 
 //
 // Delete a TagCategory.
 func (h *TagCategory) Delete(id uint) (err error) {
-	err = h.Client.Delete(Path(api.TagCategoryRoot).Inject(Params{api.ID: id}))
+	err = h.client.Delete(Path(api.TagCategoryRoot).Inject(Params{api.ID: id}))
+	return
+}
+
```
Comment (jortel) [inline binding/tagcategory.go:51]: **hint**: Coped from addon/ and generally useful enough for all binding users to be included.
Find()
Ensure()

vvv

File: binding/identity.go [hunk 9]
```diff
@@ -7,45 +7,65 @@ import (
...
-	err = h.Client.Get(api.IdentitiesRoot, &list)
+	p := Param{
+		Key:   api.Decrypted,
+		Value: "1",
+	}
+	err = h.client.Get(api.IdentitiesRoot, &list, p)
+	if err != nil {
+		return
+	}
+	for i := range list {
```
Comment (mansam) [inline binding/identity.go:50]: This conversion from resource->model->resource _appears_ to be unnecessary.

**Review by mansam — APPROVED** [review]
LGTM other than the comment about Identity.List()

File: binding/identity.go (hunk 9 above)
Comment (jortel) [reply binding/identity.go]: This is a useless left over from when With() did the decryption.
Fixed.

**Pull request merged**